cardano-testnet/test/cardano-testnet-test/Cardano/Testnet/Test/FoldBlocks.hs

```diff
@@ -72,5 +83,69 @@ prop_foldBlocks = H.integrationRetryWorkspace 2 "foldblocks" $ \tempAbsBasePath'
   -- tests that `foldBlocks` receives ledger state; once that happens,
   -- handler is called, which then writes to the `lock` and allows the
   -- test to finish.
-  _ <- H.evalIO $ IO.readMVar lock
+  _ <- H.evalIO $ H.timeout 30_000_000 $ IO.readMVar lock
   H.assert True
+
+prop_foldBlocks_check_if_treasury_is_growing :: H.Property
+prop_foldBlocks_check_if_treasury_is_growing = H.integrationRetryWorkspace 0  "foldblocks-growing-treasury" $ \tempAbsBasePath' -> do
+  -- Start testnet
+  conf@Conf{tempAbsPath=TmpAbsolutePath tempAbsPath'} <- TN.mkConf tempAbsBasePath'
+
+  let era = BabbageEra
+      options = cardanoDefaultTestnetOptions
+                  { cardanoEpochLength = 100
+                  , cardanoNodeEra = AnyCardanoEra era -- TODO: We should only support the latest era and the upcoming era
+                  , cardanoActiveSlotsCoeff = 0.3
+
+                  }
+
+  runtime@TestnetRuntime{configurationFile} <- cardanoTestnetDefault options conf
+
+  -- uncomment for epoch state live access
+  -- startLedgerNewEpochStateLogging runtime tempAbsBasePath'
+
+  -- Get socketPath
+  socketPathAbs <- do
+    socketPath' <- H.noteShowM $ H.sprocketArgumentName <$> H.headM (poolSprockets runtime)
+    H.noteIO (IO.canonicalizePath $ tempAbsPath' </> socketPath')
+
+  (_condition, treasuryValues) <- H.leftFailM . runExceptT $
+    Api.foldEpochState (File configurationFile) (Api.File socketPathAbs) Api.QuickValidation 10 M.empty handler
+  H.note_ $ "treasury for last 5 epochs: " <> show treasuryValues
+
+  let treasuriesSortedByEpoch =
+        map snd
+        . sortOn fst
+        . M.assocs
+        $ treasuryValues
+
+  if checkNonDecreasing treasuriesSortedByEpoch && checkHasIncreased treasuriesSortedByEpoch
+     then H.success
+     else do
+       H.note_ "treasury is not growing"
+       H.failure
+  where
+    handler :: AnyNewEpochState -> StateT (Map EpochNo Integer) IO LedgerStateCondition
+    handler (AnyNewEpochState _ newEpochState) = do
+      let (Coin coin) = newEpochState ^. L.nesEsL . L.esAccountStateL . L.asTreasuryL
+          epochNo = newEpochState ^. L.nesELL
+      -- handler is executed multiple times per epoch, so we keep only the latest treasury value
+      modify $ M.insert epochNo coin
+      if epochNo >= EpochNo 5
+         then pure ConditionMet
+         else pure ConditionNotMet
+
+    -- | Check if the last element > first element
+    checkHasIncreased :: (Ord a) => [a] -> Bool
+    checkHasIncreased = \case
+      [] -> False
+      x1:xs -> case reverse xs of
+                [] -> False
+                xn:_ -> xn > x1
+
+    checkNonDecreasing :: (Ord a) => [a] -> Bool
```

Comment:
Why not have this check in the handler? You would only need to check two consecutive epochs and you can do this by taking the length of the state `Map EpochNo Integer` and comparing the values.

Reply:
This seems to be more clear than combining the epoch treasury values and check in the same place.

It also checks that nothing weird is not going on - that treasury is suddenly not decreasing afterwards.

Moreover, that's a useful debug information to have how the treasury value was changing in the test period.

Reply:
Well, technically a single element set is non-decreasing. A single element set is ruled out by `checkHasIncreased`, which returns `False` for `[_]`.